zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2803,124 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a unordered list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
+            return childrenPaginated;
+        }
+
+        Set<String> childrenSet = new HashSet<>(childrenPaginated);
+        getChildrenRemainingPages(path, watcher, nextPage, childrenSet);
+
+        return new ArrayList<>(childrenSet);
+    }
+
+    private void getChildrenRemainingPages(String path,
+                                           Watcher watcher,
+                                           PaginationNextPage nextPage,
+                                           Set<String> childrenSet) throws KeeperException, InterruptedException {
+        int retryCount = 0;
+        do {
+            List<String> childrenPaginated;
+            long minCzxid = nextPage.getMinCzxid();
+            if (nextPage.getMinCzxidOffset() == 0) {
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
+                        nextPage.getMinCzxidOffset(), null, nextPage);
+            } else {
+                // If a child with the same czxid is returned in the previous page, and then deleted
+                // on the server, the children not in the previous page but offset is less than czxidOffset
+                // would be missed in the next page. Use the last child in the previous page to determine whether or not
+                // the deletion case happens. If it happens, retry fetching the page starting from offset 0.
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
+                        nextPage.getMinCzxidOffset() - 1, null, nextPage);
```

Comment:
Since we changed the offset starting from 0, do we need this -1 still?

Reply:
We still need -1 when we are fetching children starting in the middle of a transaction children (same czxid). We use the last child of the previous page to determine whether there is child deletion before the offset in the transaction children.

Only in the retry, we explicitly start from offset 0.